From 756ed1e4e2ac5feb1db7f4847099c78b0ca5082d Mon Sep 17 00:00:00 2001
From: Nev Young <neville.young@morsemicro.com>
Date: Mon, 10 Mar 2025 19:54:37 +1100
Subject: [PATCH 10/14] mac80211: introduce modparam to configure association
 timeout

Current MLME logic in mac80211 only waits one beacon interval before
failing the association. In conditions of high beacon loss triggers this
failure and requires a restart from the authentication stage. This
results in poor association times.

This change enables configuration of the association timeout by adjusting
the beacon intervals to wait for a beacon.
---
 net/mac80211/mlme.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 8d27649b3989..e7ebdf162f40 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -71,6 +71,14 @@ module_param(beacon_loss_count, int, 0644);
 MODULE_PARM_DESC(beacon_loss_count,
 		 "Number of beacon intervals before we decide beacon was lost.");
 
+/*
+ * Number of beacon intervals to wait for a beacon at association.
+ */
+static int beacon_wait_count = 1;
+module_param(beacon_wait_count, int, 0644);
+MODULE_PARM_DESC(beacon_wait_count,
+		 "Number of beacon intervals to wait for a beacon at association.");
+
 /*
  * Time the connection can be idle before we probe
  * it to see if we can still talk to the AP.
@@ -9139,11 +9147,12 @@ int ieee80211_mgd_assoc(struct ieee80211_sub_if_data *sdata,
 		if (!beacon_ies) {
 			/*
 			 * Wait up to one beacon interval ...
-			 * should this be more if we miss one?
+			 * Beacon periods to wait can be set via modparam
 			 */
 			sdata_info(sdata, "waiting for beacon from %pM\n",
 				   link->u.mgd.bssid);
-			assoc_data->timeout = TU_TO_EXP_TIME(req->bss->beacon_interval);
+			assoc_data->timeout = TU_TO_EXP_TIME(beacon_wait_count *
+				req->bss->beacon_interval);
 			assoc_data->timeout_started = true;
 			assoc_data->need_beacon = true;
 		}
-- 
2.34.1

